src/java/com/reucon/openfire/plugin/archive/xep0313/IQQueryHandler.java

```diff
@@ -278,26 +286,95 @@ private IQ buildForbiddenResponse(IQ packet) {
             }
         }
 
-        Date startDate = null;
-        Date endDate = null;
-        try {
-            if(startField != null) {
-                startDate = xmppDateTimeFormat.parseString(startField);
-            }
-            if(endField != null) {
-                endDate = xmppDateTimeFormat.parseString(endField);
-            }
-        } catch (ParseException e) {
-            Log.error("Error parsing query date filters.", e);
+       try
+        {
+	        ZonedDateTime nowDate = ZonedDateTime.now();
+	        ZonedDateTime newDate = nowDate.minusDays(conversationManager.getMaxRetrievable());
+   
+	        Date startDate = null;
+	        Date endDate = null;
+	        try {
+	        	
+	        	/*If client did not set a start field:
+	        	  - check if we have set maxRetrievable
+	        	    > yes: use the new date (now - maxRetrievable)
+	        	    > no: startDate = null for any*/
+	        	if (startField==null)
+	        	{
+	        		if (conversationManager.getMaxRetrievable()>0)
+	        		{
+	        			// we have maxRetrievable set
+	        			startDate = Date.from(newDate.toInstant());
+	        		}
+	        		else
+	        		{
+	        			// we dont have maxRetrievable set
+	        			startDate = null;
+	        		}
+	        	}
+	        	else
+	        	/*
+	        	   Client have set a start field:
+	        	     - check if we have set maxRetrievable
+	        	     > yes: check if clients date is earlier than now - maxRetrievable
+	        	     		> yes: use now - maxRetrievable as new start date
+	        	     		> no : use clients start date
+	        	     > no: use clients start date
+	        	*/
```

Comment:
Super nitpicky: but the comment now exactly duplicates the code. That is correct, but does not help to illustrate anything. Suggested improvement:

> Client has provided a start date.
> Check if the server has set a number-of-days limit to the history that clients are allowed retrieve.
> 
> - if Yes: check if the client requests data that is older than the server-defined limit
> -- if Yes: use the server-defined limit
> -- if No: use the client start date
> - if No: use the client start date